src/main/java/com/mongodb/hibernate/translate/AstVisitorValueHolder.java

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright 2024-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.hibernate.translate;
+
+import static com.mongodb.hibernate.internal.MongoAssertions.assertNotNull;
+import static com.mongodb.hibernate.internal.MongoAssertions.assertNull;
+import static com.mongodb.hibernate.internal.MongoAssertions.assertTrue;
+import static java.lang.String.format;
+
+import com.google.common.reflect.TypeToken;
```

Comment:
Does Hibernate already have a Guava dependency or would this be an additional dependency?  If the latter, I'm not sure it's wise to take it on.  What would be the alternative if we decide against it (I don't recall that the POC used it in the Attachment class)?